MdeModulePkg/Bus/Pci/PciBusDxe/PciDoe.c

```diff
@@ -19,98 +19,225 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #define PCI_VENDOR_ID_PCI_SIG       0x01
 #define PCI_DOE_PROTOCOL_DISCOVERY  0x00
 
-VOID
+#define POLL_INTERVAL_US  500
+
+/** Stalls the BSP for the minimum of POLL_INTERVAL_US and Timeout.
+
+   @param[in]  Timeout    The time limit in microseconds remaining for
+                          the DOE operation.
+
+   @retval     StallTime  Time of execution stall.
+**/
+STATIC
+UINTN
+CalculateAndStallInterval (
+  IN UINTN  Timeout
+  )
+{
+  UINTN  StallTime;
+
+  if ((Timeout < POLL_INTERVAL_US) && (Timeout != 0)) {
+    StallTime = Timeout;
+  } else {
+    StallTime = POLL_INTERVAL_US;
+  }
+
+  gBS->Stall (StallTime);
+
+  return StallTime;
+}
+
+/** Wait for the DOE Not Busy bit to be set
+
+  @param[in]  PciIoDevice    A pointer to the EFI_PCI_IO_PROTOCOL instance.
+
+  @retval EFI_SUCCESS Source was updated to support Handler.
+  @retval EFI_UNSUPPORTED  Hardware could not be programmed.
+  @retval EFI_TIMEOUT  Did not complete in timeout period.
+**/
+EFI_STATUS
 DoeWaitForNotBusy (
   IN PCI_IO_DEVICE  *PciIoDevice
   )
 {
+  EFI_STATUS                  Status;
   PCI_EXPRESS_REG_DOE_STATUS  DoeStatus;
+  UINTN                       Timeout;
 
   //
   // Read DOE Status
   //
-  PciIoDevice->PciIo.Pci.Read (
-                           &PciIoDevice->PciIo,
-                           EfiPciIoWidthUint32,
-                           PciIoDevice->DoeCapabilityOffset + PCI_EXPRESS_REG_DOE_STATUS_OFFSET,
-                           1,
-                           &DoeStatus
-                           );
+  Status = PciIoDevice->PciIo.Pci.Read (
+                                    &PciIoDevice->PciIo,
+                                    EfiPciIoWidthUint32,
+                                    PciIoDevice->DoeCapabilityOffset + PCI_EXPRESS_REG_DOE_STATUS_OFFSET,
+                                    1,
+                                    &DoeStatus
+                                    );
+  if (EFI_ERROR (Status)) {
+    return EFI_UNSUPPORTED;
+  }
+
+  //
+  // Section "6.30.2 Operation" describes a range of operations
+  // that must complete within 1 second. The device could be busy
+  // for any number of reasons, so let's use a 1 second
+  // timeout to wait a reasonable amount of time, without stalling
+  // if an error occurs.
+  //
+  // Timeout is in microseconds
+  //
+  Timeout = 1000000;

```

Comment:
Recommend using #define for this value with the comment and spec reference with the #define.  I see this same value and spec reference in more than one location.

Reply:
Done!